src/common.cpp

```diff
@@ -11,20 +11,23 @@ common::ThreadPool::ThreadPool(size_t n_threads,
                                std::function<void(const std::string&)> err)
   : error_handler_{ err }
 {
-    if (n_threads == 0) {
-        throw std::runtime_error("Cannot create thread pool with 0 threads.");
-    }
-
-    for (auto i = 0; i < n_threads; ++i) {
-        threads_.emplace_back([this] { thread_worker_(); });
-    }
+    n_threads_ = std::clamp(
+      n_threads, (size_t)1, (size_t)std::thread::hardware_concurrency());
```

Comment:
It seems possible that [`hardware_concurrency` can return 0](https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency) and the [behavior of `clamp` is undefined when `lo > hi`](https://en.cppreference.com/w/cpp/algorithm/clamp).

So consider guarding against that (likely rare) case, by checking the return value of `hardware_concurrency` and/or using `max` of it and 1.